includes/student-progress/course-progress/repositories/class-aggregate-course-progress-repository.php

```diff
@@ -0,0 +1,120 @@
+<?php
+/**
+ * File containing the class Aggregate_Course_Progress_Repository.
+ *
+ * @package sensei
+ */
+
+namespace Sensei\Student_Progress\Course_Progress\Repositories;
+
+use Sensei\Student_Progress\Course_Progress\Models\Course_Progress;
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit;
+}
+
+/**
+ * Class Aggregate_Course_Progress_Repository.
+ *
+ * Aggregate repository is an intermediate repository that delegates the calls to the appropriate repository implementation.
+ *
+ * @since $$next-version$$
+ */
+class Aggregate_Course_Progress_Repository implements Course_Progress_Repository_Interface {
+	/**
+	 * Comments based course progress repository implementation.
+	 *
+	 * @var Comments_Based_Course_Progress_Repository
+	 */
+	private $comments_based_repository;
+
+	/**
+	 * Tables based course progress repository implementation.
+	 *
+	 * @var Tables_Based_Course_Progress_Repository
+	 */
+	private $tables_based_repository;
+
+	/**
+	 * The flag if the tables based implementation is available for use.
+	 *
+	 * @var bool
+	 */
+	private $use_tables;
+
+	/**
+	 * Aggregate_Course_Progress_Repository constructor.
+	 *
+	 * @param Comments_Based_Course_Progress_Repository $comments_based_repository Comments based course progress repository implementation.
+	 * @param Tables_Based_Course_Progress_Repository   $tables_based_repository  Tables based course progress repository implementation.
+	 * @param bool                                      $use_tables  The flag if the tables based implementation is available for use.
+	 */
+	public function __construct( Comments_Based_Course_Progress_Repository $comments_based_repository, Tables_Based_Course_Progress_Repository $tables_based_repository, bool $use_tables ) {
+		$this->comments_based_repository = $comments_based_repository;
+		$this->tables_based_repository   = $tables_based_repository;
+		$this->use_tables                = $use_tables;
+	}
+
+	/**
+	 * Creates a new course progress.
+	 *
+	 * @param int $course_id The course ID.
+	 * @param int $user_id The user ID.
+	 * @return Course_Progress The course progress.
+	 */
+	public function create( int $course_id, int $user_id ): Course_Progress {
+		$progress = $this->comments_based_repository->create( $course_id, $user_id );
+		if ( $this->use_tables ) {
+			$this->tables_based_repository->create( $course_id, $user_id );
+		}
+		return $progress;
+	}
+
+	/**
+	 * Gets a course progress.
+	 *
+	 * @param int $course_id The course ID.
+	 * @param int $user_id The user ID.
+	 * @return Course_Progress|null The course progress or null if it does not exist.
+	 */
+	public function get( int $course_id, int $user_id ): ?Course_Progress {
+		return $this->comments_based_repository->get( $course_id, $user_id );
```

Comment:
Looks like we should create a tables based course here in case we don't have it yet (same for lessons and quizzes).

Reply:
This is because we always need to return a progress instance? In that case, does it make sense to rename the method to `get_or_create`?

I'm not completely sure if I understand the problem. Could you give more context?

Reply:
Hey 👋 No, my comment was about the data inconsistency: we have an entry for the progress in comments, but don't have it in tables.
To get a kind of eventual consistency, we need to create that entry in tables.
Like that:
```
progress = this->comments->get(course, user);
if (progress) {
  if (!this->tables->has(course, user)) {
    tables_progress = this->tables->create(course, user)
    tables_progress_to_save = Progress(tables_progress->id, ...); // set fields from comments progress
    this->tables->save(tables_progress_to_save)
  }
}
return progress
```
At some point, doing that way, we get the same state of the progress in comments and tables.
But right now, I think we don't need it as we rely on migration. This approach won't give us the consistency immediately, and for the cold data we won't get it ever.
So, that was a bad idea :) We just have to accept that we have some inconsistency until we run data migration.

Reply:
Got it, thanks.

> We just have to accept that we have some inconsistency until we run data migration.

I think that's expected. 👍